examples/sockeye_translate/README.md

```diff
@@ -0,0 +1,71 @@
+# sockeye-serving
+This example shows how to serve Sockeye models for machine translation.
+The custom handler is implemented in `sockeye_service.py`.
+Since Sockeye has many dependencies, it's convenient to use Docker.
+For simplicity, we'll use a pre-trained model and make some assumptions about how we preprocess the data.
+
+## Getting Started With Docker
+Pull the latest Docker image:
+```bash
+docker pull jwoo11/sockeye-serving
+```
+
+Download the example model archive file (MAR).
+This is a ZIP archive containing the parameter files and scripts needed to run translation for a particular language:
+* https://www.dropbox.com/s/pk7hmp7a5zjcfcj/zh.mar?dl=0
+
+Extract the MAR file to `/tmp/models`.
+ We'll use this directory as a bind mount for Docker:
+```bash
+unzip -d /tmp/models/zh zh.mar
+```
+
+Start the server:
+```bash
+docker run -itd --name mms -p 8080:8080 -p 8081:8081 -v /tmp/models/:/models jwoo11/sockeye-serving serve
+```
+
+Now we can load the model using the management API provided by `mxnet-model-server`:
```

Comment:
If you extracted model into /opt/ml/model, and use default config.properties file, you don't need load model,
the model will be loaded at startup time:

model_store=/opt/ml/model
load_models=ALL

Reply:
Getting the following with the above config: 
> Invalid configuration: Model zh is already registered.
The logs didn't provide much detail.

Reply:
Do you get this error when you load the model with the API or do you get the error when the model-server is started, with ```docker --run``` command?  You wouldn't need to run the ```curl -X POST "http://localhost:8081/models?synchronous=true&initial_workers=1&url=zh"
``` command if you put files in the ```/opt/ml/model``` folder and you have the above config in your ```config.properties```. With the above configuration, all the models in ```/opt/ml/model``` would be loaded onto the model-server at boot-up time.

Reply:
Turns out I had a duplicate model sitting in the same directory--the unzipped version of the MAR. When I got rid of the unzipped folder my custom handler couldn't find the path to a script. So I tried using just the unzipped folder and my Docker daemon crashed. I think all CPU cores were being utilized and then Docker became unresponsive. I see repeated attempts to restart workers in the logs:

`2019-01-28 16:53:05,821 [WARN ] W-9005-zh com.amazonaws.ml.mms.wlm.WorkerThread - Backend worker thread exception.
java.lang.IllegalMonitorStateException
	at java.util.concurrent.locks.ReentrantLock$Sync.tryRelease(ReentrantLock.java:151)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer.release(AbstractQueuedSynchronizer.java:1261)
	at java.util.concurrent.locks.ReentrantLock.unlock(ReentrantLock.java:457)
	at com.amazonaws.ml.mms.wlm.Model.pollBatch(Model.java:173)
	at com.amazonaws.ml.mms.wlm.BatchAggregator.getRequest(BatchAggregator.java:45)
	at com.amazonaws.ml.mms.wlm.WorkerThread.run(WorkerThread.java:133)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
2019-01-28 16:53:05,828 [DEBUG] W-9005-zh com.amazonaws.ml.mms.wlm.WorkerThread - W-9005-zh State change WORKER_MODEL_LOADED -> WORKER_STOPPED
2019-01-28 16:53:05,829 [INFO ] W-9005-zh com.amazonaws.ml.mms.wlm.WorkerThread - Retry worker: 9005 in 1 seconds.
2019-01-28 16:53:07,092 [WARN ] W-9002-zh-stderr com.amazonaws.ml.mms.wlm.WorkerLifeCycle - [INFO:sockeye.model] Loaded params from "/opt/ml/model/zh/params.best"
2019-01-28 16:53:07,297 [WARN ] W-9002-zh-stderr com.amazonaws.ml.mms.wlm.WorkerLifeCycle - [INFO:sockeye.inference] 1 model(s) loaded in 4.3263s
2019-01-28 16:53:10,253 [INFO ] epollEventLoopGroup-4-3 com.amazonaws.ml.mms.wlm.WorkerThread - 9002 Worker disconnected. WORKER_STARTED
2019-01-28 16:53:10,261 [DEBUG] W-9002-zh com.amazonaws.ml.mms.wlm.WorkerThread - Backend worker monitoring thread interrupted or backend worker process died.
java.lang.InterruptedException
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.reportInterruptAfterWait(AbstractQueuedSynchronizer.java:2014)
	at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.awaitNanos(AbstractQueuedSynchronizer.java:2088)
	at java.util.concurrent.ArrayBlockingQueue.poll(ArrayBlockingQueue.java:418)
	at com.amazonaws.ml.mms.wlm.WorkerThread.run(WorkerThread.java:139)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
2019-01-28 16:53:10,271 [WARN ] W-9002-zh com.amazonaws.ml.mms.wlm.BatchAggregator - Load model failed: zh, error: Worker died`

Reply:
@jamesewoo 
This looks like a bug to me. The failure should happen at registration time, not load model time. And it should not retry.

Reply:
With `load_models=ALL` I get 6 CPU worker instances on a 6 core machine, but I'm unable to get translations back. If I leave that config property out and scale it manually, translation works with up to 4 workers but fails with 5 or 6:
`{
  "code": 404,
  "type": "ModelNotFoundException",
  "message": "Load failed... Deregistered model zh"
}`

Reply:
@jamesewoo 
Can you share your logs?

By the way, you can limit number of worker that loaded on startup in config.properties:
default_workers_per_model=1

Reply:
Logs for 5 initial workers: https://pastebin.com/5eBJehEp
I think it has something to do with how I'm passing device IDs from MMS to Sockeye. I got the restarting to stop, but the error is still "model not found."

Reply:
I think whats happening is this. The model-server is trying to register the model in synchronous mode. The model itself is taking more than 2 mins to load. This causes the timeout to occur and that kills the backend workers. Since the register call was synchronous and the creation of workers failed, the model is being unregistered. So, you wouldn't be able to query for the model because its already unregistered. You will get a "model not found" exception. So, the key culprit seems to be the time taken to load the model. Could that be an issue?

Reply:
I don't think it's a timeout issue--the model loads into 4 workers within a few seconds. It's definitely failing to load for more than 4 workers for some reason. Should I be using the `gpu_id` property from the context if I only want to use CPUs?

Reply:
I read that MXNet takes all available CPUs, so the device ID passed to `mx.cpu()` doesn't matter. Still can't get more than 4 workers to load. Should we create a new issue for this problem?